src/urllib3/response.py

```diff
@@ -876,39 +898,72 @@ def read(
 
         data = self._raw_read(amt)
 
-        flush_decoder = amt is None or (amt != 0 and not data)
-
         if not data and len(self._decoded_buffer) == 0:
             return data
 
         if amt is None:
-            data = self._decode(data, decode_content, flush_decoder)
+            data = self._decode(data, decode_content, flush_decoder=True)
             if cache_content:
                 self._body = data
         else:
             # do not waste memory on buffer when not decoding
             if not decode_content:
-                if self._has_decoded_content:
-                    raise RuntimeError(
-                        "Calling read(decode_content=False) is not supported after "
-                        "read(decode_content=True) was called."
-                    )
                 return data
 
+            flush_decoder = not (data or amt == 0)
             decoded_data = self._decode(data, decode_content, flush_decoder)
             self._decoded_buffer.put(decoded_data)
 
             while len(self._decoded_buffer) < amt and data:
-                # TODO make sure to initially read enough data to get past the headers
-                # For example, the GZ file header takes 10 bytes, we don't want to read
-                # it one byte at a time
-                data = self._raw_read(amt)
+                data = self._raw_read(8192, one_read=True)
+                flush_decoder = not data
                 decoded_data = self._decode(data, decode_content, flush_decoder)
                 self._decoded_buffer.put(decoded_data)
             data = self._decoded_buffer.get(amt)
 
         return data
 
+    def read1(
+        self,
+        amt: int | None = None,
+        decode_content: bool | None = None,
+    ) -> bytes:
+        self._init_decoder()
+        if decode_content is None:
+            decode_content = self.decode_content
+
+        # seems unlikely that 32MiB will ever be available in the OS buffer
+        if amt is None:
+            amt = 2**25
+
+        if len(self._decoded_buffer) >= amt:
+            return self._decoded_buffer.get(amt)
```

Comment:
One `SSL_read` call [can return 16kB](https://www.openssl.org/docs/man3.2/man3/SSL_read.html#:~:text=The%20data%20are%20received%20in%20records%20(with%20a%20maximum%20record%20size%20of%2016kB).) of data; a small experiment shows that this amount may be decompressed to much more than 32MiB.
```sh
$ dd if=/dev/zero bs=64MiB count=1 | tr '\000' '1'  > 64MiB.txt
1+0 records in
1+0 records out
67108864 bytes (67 MB, 64 MiB) copied, 0.138876 s, 483 MB/s
$ du -sh 64MiB.txt   
64M	64MiB.txt
$ brotli 64MiB.txt
$ du -sh 64MiB.txt.br 
4.0K	64MiB.txt.br
```

Is there a reason not to return everything from `self._decoded_buffer` and make no `self._raw_read` calls when `self._decoded_buffer` is not empty?

Reply:
That's an impressive expansion factor!  My use case didn't want enormous blocks of memory being pushed around unnecessarily, but don't think I thought it through.  They'd still exist in this case, just buffered (i.e. in `_decoded_buffer`), and hence it makes sense to just return them here when a caller asks for them.

Will update the code to actually return everything when it was requested!

Am still a bit worried about running out of memory (i.e. attempting to allocate very large `bytes` buffers for returning things like this), are there other bits of code you know of that deal with this that you know of?